### PR TITLE
Reworked Password-Reset flow

### LINKED ISSUE #81: [Bug] Don't send initial admin password via email, or any password for that matter

### PR DESCRIPTION
**Pull request by Anuril**
The reset password Functionality has been reworked. 

If a user presses the Reset Password button, they will get an E-Mail with a link to a page to reset their password.

The hash in the Link is only generated once (Previously it was generated each time someone clicked). If the Button is pressed again, the same link will be sent again, but only if the last E-Mail has been sent more than 60 seconds prior.

When the user changes their password, they will receive an E-Mail with the information that their password has been changed. (Without login information in it of course.)

The redirect now points to the correct login page.

TLDR;
- No more (user & admin) Passwords sent via E-Mail
- Fixed possible DoS
- Correct redirects

### CODE REVIEW

**Review by jaapmarcus — APPROVED** [review]

File: src/modules/Client/Controller/Client.php [hunk 1]
```diff
@@ -63,15 +63,21 @@ public function get_client_page(\Box_App $app, $page)
...
+        $service = $this->di['mod_service']('client');
         $this->di['events_manager']->fire(['event' => 'onBeforePasswordResetClient']);
         $data = [
             'hash' => $hash,
         ];
-        $api->client_confirm_reset($data);
-        $app->redirect('/login');
+        $template = 'mod_client_set_new_password';
+        
+        // Call pwreset_valid API and if true, then render the template, otherwise redirect to the index page
```
Comment (BelleNottelling) [inline src/modules/Client/Controller/Client.php:75]: > Call pwreset_valid API

It's not an API endpoint

File: src/modules/Client/Service.php [hunk 2]
```diff
@@ -680,4 +681,51 @@ public function exportCSV(array $headers)
...
         return $this->di['table_export_csv']('client', 'clients.csv', $headers);
     }
+    /**
+     * Confirm password reset action.
+     *
+     * @return bool
+     *
+     * @throws \Box_Exception
+     */
+    public function pwreset_valid($data)
```
Comment (BelleNottelling) [inline src/modules/Client/Service.php:691]: I don't like shortening the name of this function

File: src/modules/Client/Service.php [hunk 3]
```diff
@@ -680,4 +681,51 @@ public function exportCSV(array $headers)
...
+    {
+        $required = [
+            'hash' => 'Hash required',
+        ];
+        $this->di['events_manager']->fire(['event' => 'onBeforePasswordResetClient']);
+        $this->di['validator']->checkRequiredParamsForArray($required, $data);
+
+        $reset = $this->di['db']->findOne('ClientPasswordReset', 'hash = ?', [$data['hash']]);
+        if (!$reset instanceof \Model_ClientPasswordReset) {
+            throw new \Box_Exception('The link have expired or you have already confirmed password reset.');
```
Comment (BelleNottelling) [inline src/modules/Client/Service.php:701]: You fixed the grammatical error in this exception elsewhere, please use the fixed version here

File: src/modules/Client/Api/Guest.php [hunk 4]
```diff
@@ -150,78 +151,99 @@ public function login($data)
...
     {
         $required = [
-            'hash' => 'Hash required',
+            'hash' => 'No Hash provided',
+            'password' => 'Password required',
+            'password_confirm' => 'Password confirmation required',
         ];
-        $this->di['events_manager']->fire(['event' => 'onBeforePasswordResetClient']);
-        $this->di['validator']->checkRequiredParamsForArray($required, $data);
+        $this->di['events_manager']->fire(['event' => 'onBeforeClientProfilePasswordSet', 'params' => $data['hash']]);
```
Comment (BelleNottelling) [inline src/modules/Client/Api/Guest.php:214]: This should really be `onBeforeClientProfilePasswordChange` (an already existing event that's in use & the "after" variant is already being sent at this end of this function)

**Review by BelleNottelling — CHANGES_REQUESTED** [review]
Please see the comments I made

File: src/modules/Client/Service.php [hunk 5]
```diff
@@ -680,4 +681,51 @@ public function exportCSV(array $headers)
...
+            return false;
+        } else {
+            return $c->id;
+        }
+    }
+
+    /*
+     * Function that cleans out expired password reset requests ( 15 minutes )
+     * 
+     * @return bool
```
[the remark this answers is not in the document]
Comment (Anuril) [reply src/modules/Client/Service.php:717]: Done

File: src/modules/Client/Api/Guest.php [hunk 6]
```diff
@@ -150,78 +151,99 @@ public function login($data)
...
+        // prepare reset email
+        $email = [
+            'to_client' => $c->id,
+            'code' => 'mod_client_password_reset_request',
+            'hash' => $reset->hash,
+        ];
+        // Send the email if the reset request has the same created_at and updated_at
         $emailService = $this->di['mod_service']('email');
-        $emailService->sendTemplate($email);
+        if ($reset->created_at == $reset->updated_at) {
```
[the remark this answers is not in the document]
Comment (Anuril) [reply src/modules/Client/Api/Guest.php:192]: If the reset object has just been created, we need to send the email. This can be checked by making sure created at and updated at are the same. 

If we do not send the email in this condition, the email is never sent until the user requests it again - minimum 1 minute later. 

We're also not allowing new password resets to be requested, only to be re-sent. 

You're right, the send_now was missing, I'll fix that :)

[the remark this answers is not in the document]
Comment (BelleNottelling) [reply src/modules/Client/Api/Guest.php:192]: Okay, that makes sense to me now.
I was thinking about it wrong, thanks!

File: src/modules/Client/Api/Guest.php (hunk 4 above)
Comment (Anuril) [reply src/modules/Client/Api/Guest.php:214]: I had that before, but then I though - maybe we _don't_ want to use the same event, because ```onBeforeClientProfilePasswordChange``` is used when the user manually changes their password from a logged in state, and ```onBeforeClientProfilePasswordSet``` happens when the user resets their password. Maybe we want to differentiate between these two events for auditing purposes?

Comment (BelleNottelling) [reply src/modules/Client/Api/Guest.php:214]: I'm indifferent to the two, but I think `onBeforeClientProfilePasswordReset` makes more sense if we are to implement a new one.
And if we are going to implement one, we should fire an "after" variant of it as well (It's not technically required, just generally how everything else is done and I'd like to keep it consistent)

Comment (Anuril) [reply src/modules/Client/Api/Guest.php]: I've renamed it.

**Review by BelleNottelling — APPROVED** [review]
